test/feature/dynamic_types/DynamicTypesTests.cpp

```diff
@@ -0,0 +1,7712 @@
+    ASSERT_EQ(RETCODE_OK, struct_data->get_int64_value(test4, 1));
+    ASSERT_EQ(0, test4);
+    ASSERT_EQ(RETCODE_OK, struct_data->get_int32_value(test2, 2));
+    ASSERT_EQ(test1, test2);
+    ASSERT_EQ(RETCODE_OK, struct_data->get_string_value(test6, 4));
+    ASSERT_EQ("", test6);
+
+    DynamicDataFactory::get_instance()->delete_data(struct_data);
+}
+
+TEST_F(DynamicTypesTests, DynamicType_multi_structure)
+{
+    DynamicTypeBuilderFactory::_ref_type factory {DynamicTypeBuilderFactory::get_instance()};
+
+    TypeDescriptor::_ref_type type_descriptor {traits<TypeDescriptor>::make_shared()};
+    type_descriptor->kind(TK_STRUCTURE);
+    type_descriptor->name("InnerStructTest");
+    DynamicTypeBuilder::_ref_type builder {factory->create_type(type_descriptor)};
+    ASSERT_TRUE(builder);
+
+    // Add members to the struct.
+    MemberDescriptor::_ref_type member_descriptor {traits<MemberDescriptor>::make_shared()};
+    member_descriptor->type(factory->get_primitive_type(TK_INT32));
+    member_descriptor->name("int32");
+    member_descriptor->id(0);
+    ASSERT_EQ(builder->add_member(member_descriptor), RETCODE_OK);
+
+    member_descriptor = traits<MemberDescriptor>::make_shared();
+    member_descriptor->type(factory->get_primitive_type(TK_INT64));
+    member_descriptor->name("int64");
+    member_descriptor->id(1);
+    ASSERT_EQ(builder->add_member(member_descriptor), RETCODE_OK);
+
+    DynamicType::_ref_type inner_struct_type {builder->build()};
+    ASSERT_TRUE(inner_struct_type);
+
+    // Create the parent struct.
+    type_descriptor = traits<TypeDescriptor>::make_shared();
+    type_descriptor->kind(TK_STRUCTURE);
+    type_descriptor->name("StructTest");
+    builder = factory->create_type(type_descriptor);
+    ASSERT_TRUE(builder);
+
+    // Add members to the struct.
+    member_descriptor = traits<MemberDescriptor>::make_shared();
+    member_descriptor->type(inner_struct_type);
+    member_descriptor->name("Structure");
+    member_descriptor->id(0);
+    ASSERT_EQ(builder->add_member(member_descriptor), RETCODE_OK);
+
+    member_descriptor = traits<MemberDescriptor>::make_shared();
+    member_descriptor->type(factory->get_primitive_type(TK_INT64));
+    member_descriptor->name("int64");
+    member_descriptor->id(10);
+    ASSERT_EQ(builder->add_member(member_descriptor), RETCODE_OK);
+
+    DynamicType::_ref_type struct_type {builder->build()};
+    ASSERT_TRUE(struct_type);
+
+    DynamicData::_ref_type struct_data {DynamicDataFactory::get_instance()->create_data(struct_type)};
+    ASSERT_TRUE(struct_data);
+
+    // Test get_member_by_name and get_member_by_index.
+    ASSERT_EQ(MEMBER_ID_INVALID, struct_data->get_member_id_by_name(""));
+    ASSERT_EQ(0, struct_data->get_member_id_by_name("Structure"));
+    ASSERT_EQ(10, struct_data->get_member_id_by_name("int64"));
+    ASSERT_EQ(0, struct_data->get_member_id_at_index(0));
+    ASSERT_EQ(10, struct_data->get_member_id_at_index(1));
+  
```

Comment:
I think this is not correct. Getting the complex value for a non-selected union member fails. I would expect the same to happen for loans. And if we decide that the loan can be returned and if set, the discriminator points to the new union member, then the same philosophy must be followed with complex API.

Also, the test only uses the get API and this should fail because the union member being read is not the one selected.

Reply:
We discuss this again and the result it is correct now.